File: backend/call-service/app/schemas/calls.py
from pydantic import BaseModel, Field


class IceServer(BaseModel):
    urls: list[str] | str
    username: str | None = None
    credential: str | None = None


class IceConfigResponse(BaseModel):
    ice_servers: list[IceServer]


class CreateCallRequest(BaseModel):
    call_type: str = Field(pattern="^(audio|video)$")
    participant_ids: list[str] = Field(min_length=1)
    conversation_id: str | None = None


class CallResponse(BaseModel):
    id: str
    call_type: str
    status: str
    caller_id: str
    participant_ids: list[str]
    conversation_id: str | None
    is_group: bool
    # "mesh" = peer-to-peer (1:1); "sfu" = join the LiveKit room with a token.
    mode: str = "mesh"
    created_at: str


class SignalRequest(BaseModel):
    to_user_id: str
    signal_type: str = Field(description="offer|answer|ice|hangup|screen")
    sdp: dict | None = None
    candidate: dict | None = None


class CallTokenResponse(BaseModel):
    """Everything a client needs to join the SFU room for a group call."""

    room: str
    url: str
    token: str
    expires_in: int

File: frontend/web/src/api/calls.ts
import { apiFetch } from "./client";

export type CallType = "audio" | "video";

export interface IceServer {
  urls: string | string[];
  username?: string;
  credential?: string;
}

export interface CallSession {
  id: string;
  call_type: CallType;
  status: string;
  caller_id: string;
  participant_ids: string[];
  conversation_id: string | null;
  is_group: boolean;
  created_at: string;
}

export async function listCalls(): Promise<CallSession[]> {
  return apiFetch<CallSession[]>("/calls/calls");
}

export async function getIceConfig(): Promise<{ ice_servers: IceServer[] }> {
  return apiFetch("/calls/ice");
}

export async function createCall(body: {
  call_type: CallType;
  participant_ids: string[];
  conversation_id?: string;
}): Promise<CallSession> {
  return apiFetch<CallSession>("/calls/calls", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function acceptCall(callId: string): Promise<CallSession> {
  return apiFetch<CallSession>(`/calls/calls/${callId}/accept`, { method: "POST", body: "{}" });
}

export async function rejectCall(callId: string): Promise<void> {
  await apiFetch(`/calls/calls/${callId}/reject`, { method: "POST", body: "{}" });
}

export async function endCall(callId: string): Promise<void> {
  await apiFetch(`/calls/calls/${callId}/end`, { method: "POST", body: "{}" });
}

export async function sendCallSignal(
  callId: string,
  body: {
    to_user_id: string;
    signal_type: string;
    sdp?: RTCSessionDescriptionInit;
    candidate?: RTCIceCandidateInit;
  },
): Promise<void> {
  await apiFetch(`/calls/calls/${callId}/signal`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}
